File: js/05.js
var fs = require("fs")

input = fs.readFileSync("../input/05-01.txt", "utf-8")

res = input
  .split("\n")
  .filter((x) => x.length > 1)
  .map((line) => {

    lb = 0
    ub = 127

    for (part of line.substr(0, 7)) {
      amt = Math.floor((ub - lb)/2) + 1
      if (part === "F") {
        ub -= amt
      } else {
        lb += amt
      }
    }

    row = (line.charAt(6) === "F") ? lb : ub

    lb = 0
    ub = 7

    for (part of line.substr(7, 3)) {
      amt = Math.floor((ub - lb)/2) + 1
      if (part === "L") {
        ub -= amt
      } else {
        lb += amt
      }
    }

    seat = (line.charAt(9) === "R") ? lb : ub

    return((row * 8) + seat)

  }).sort()

// 05-01

console.log(Math.max(...res))

// 05-02

diff = res.map((val, idx, arr) => idx ? val - arr[idx-1] : 0 - val)

console.log(res[diff.findIndex((x) => x > 1)] - 1)

File: R/05.py
import math
import numpy as np  

with open("../input/05-01.txt") as f:
  input = f.read().splitlines()

def partition(seq, lb, ub, trigger):
  
  for idx, part in enumerate(seq):
    amt = math.floor((ub - lb)/2) + 1
    if (part == trigger):
      ub -= amt
    else:
      lb += amt
  
  return(ub if (seq[-1] == trigger) else lb)


def get_seat(x):
  return(
    (partition(x[0:7],  0, 127, "F") * 8) + 
     partition(x[7:10], 0,   7, "L") 
  )

res = [ get_seat(bpass) for bpass in input ]

# 05-01

res.sort()

max(res)

# 05-02

idx = np.where(np.diff(res) > 1)

res[idx[0][0]] + 1

